packages/regl-worldview/src/WorldviewContext.js

```diff
@@ -240,63 +244,104 @@ export class WorldviewContext {
         // clear the framebuffer
         regl.clear({ color: [0, 0, 0, 1], depth: 1 });
+        let currentObjectId = 0;
+        const seenObjects = [];
+        const seenIds: Array<HitmapId> = [];
+        let counter = 0;
 
         // draw the hitmap components to the framebuffer
+
         camera.draw(this.cameraStore.state, () => {
-          this._drawInput(true);
-
-          let objectId = 0;
-
-          // it's possible to get x/y values outside the framebuffer size
-          // if the mouse quickly leaves the draw area during a read operation
-          // reading outside the bounds of the framebuffer causes errors
-          // and puts regl into a bad internal state.
-          // https://github.com/regl-project/regl/blob/28fbf71c871498c608d9ec741d47e34d44af0eb5/lib/read.js#L57
-          if (x < Math.floor(width) && y < Math.floor(height) && x >= 0 && y >= 0) {
-            const pixel = new Uint8Array(4);
-
-            // read pixel value from the frame buffer
-            regl.read({
-              x,
-              y,
-              width: 1,
-              height: 1,
-              data: pixel,
-            });
-
-            objectId = getIdFromColor(pixel);
-          }
-          resolve(objectId);
+          do {
+            if (counter === MAX_NUMBER_OF_LAYERS) {
+              // Provide a max number of layers so this while loop doesn't crash the page.
+              console.error(
+                `Hit ${MAX_NUMBER_OF_LAYERS} iterations. There is either that number of rendered layers or a bug with hitmap events.`
+              );
+              break;
+            } else if (counter > 0 && !enableStackedObjectEvents) {
+              // Only allow multiple object selection if we've explicitly enabled it.
+              break;
+            }
+            counter++;
+            regl.clear({ color: [0, 0, 0, 1], depth: 1 });
+            this._drawInput(true, seenObjects);
+
+            // it's possible to get x/y values outside the framebuffer size
+            // if the mouse quickly leaves the draw area during a read operation
+            // reading outside the bounds of the framebuffer causes errors
+            // and puts regl into a bad internal state.
+            // https://github.com/regl-project/regl/blob/28fbf71c871498c608d9ec741d47e34d44af0eb5/lib/read.js#L57
+            if (x < Math.floor(width) && y < Math.floor(height) && x >= 0 && y >= 0) {
+              const pixel = new Uint8Array(4);
+
+              // read pixel value from the frame buffer
+              regl.read({
+                x,
+                y,
+                width: 1,
+                height: 1,
+                data: pixel,
+              });
+
+              currentObjectId = getIdFromColor(pixel);
+              if (currentObjectId && this.getDrawPropByHitmapId(currentObjectId).object) {
+                seenIds.push(currentObjectId);
+                seenObjects.push(this.getDrawPropByHitmapId(currentObjectId));
```

Comment:
maybe use a Set instead of Array for these?

Reply:
Unfortunately we can't use a set, because we would need access to the original mouse event object. Instead we're reading from the sub-object inside of it and the instance index separately, and there isn't really an elegant way to get around this. The total number of seen objects should be <10 in the common case anyways